File: backend/app/schemas/payment.py
"""支付订单请求和只读响应模型。"""
from datetime import datetime
from decimal import Decimal
from pydantic import BaseModel, ConfigDict, Field
from app.services.payment_provider import PaymentMethod


class PaymentCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")
    booking_id: int
    payment_method: PaymentMethod = PaymentMethod.CARD_CHECKOUT


class PaymentMethodAvailability(BaseModel):
    method: PaymentMethod
    available: bool
    test_mode: bool
    reason: str | None = None


class PaymentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str; order_id: str; payment_attempt_id: str; booking_id: int; user_id: int; amount: int
    status: str; payment_method: str; provider: str
    order_status: str | None = None
    settlement_currency: str; settlement_amount_minor: int
    cny_reference_amount_minor: int; property_currency: str
    exchange_rate: Decimal; exchange_rate_source: str
    exchange_rate_timestamp: datetime; expires_at: datetime
    checkout_url: str | None = None; snapshot: dict = Field(default_factory=dict)
    transaction_id: str | None = None; paid_at: datetime | None = None
    created_at: datetime; updated_at: datetime


class PaymentResultResponse(PaymentResponse):
    """支付结果页所需的后端可信订单摘要。"""
    property_image_url: str | None = None
    booking_created_at: datetime
    status_updated_at: datetime
    failure_reason: str | None = None


class TenantOrderListItem(BaseModel):
    booking_id: int
    order_id: str
    agreement_id: str
    agreement_number: str
    property_id: int
    property_name: str
    property_image_url: str | None = None
    property_city: str
    property_address: str
    lease_start_date: str | None = None
    lease_end_date: str | None = None
    lease_months: int | None = None
    settlement_currency: str
    settlement_amount_minor: int
    cny_reference_amount_minor: int
    property_currency: str
    property_amount_minor: int
    order_status: str
    payment_status: str
    booking_status: str
    status_label: str
    created_at: datetime
    expires_at: datetime
    remaining_payment_seconds: int
    can_pay: bool
    payment_action_label: str | None = None
    failure_reason: str | None = None


class TenantOrderDetail(TenantOrderListItem):
    applicant_name: str
    applicant_phone_masked: str | None = None
    applicant_email_masked: str | None = None
    property_type: str
    property_country: str
    property_description: str | None = None
    monthly_rent_minor: int
    deposit_amount_minor: int
    service_fee_amount_minor: int
    tax_amount_minor: int
    exchange_rate: Decimal
    exchange_rate_source: str
    exchange_rate_timestamp: datetime
    status_updated_at: datetime
    paid_at: datetime | None = None
    transaction_id_masked: str | None = None
    webhook_confirmed: bool = False
    amounts_verified: bool = False
    inventory_reserved: bool = False


class PaymentEligibilityResponse(BaseModel):
    booking_id: int
    can_pay: bool
    order_status: str
    payment_status: str
    expires_at: datetime
    reason: str | None = None
    payment_id: str | None = None


class TenantOrderListResponse(BaseModel):
    items: list[TenantOrderListItem]
    total: int

File: frontend/src/services/payment.ts
// 支付 API：金额和汇率只读取服务端支付快照。
import api from './api'

export interface Money { currency: string; minor_units: number; minor_unit_exponent: number; decimal: string }
export interface PaymentResponse {
  id: string; order_id: string; payment_attempt_id: string; booking_id: number; user_id: number; status: string; order_status: string | null; provider: string
  payment_method: string; settlement_currency: string; settlement_amount_minor: number
  cny_reference_amount_minor: number; property_currency: string; exchange_rate: string
  exchange_rate_source: string; exchange_rate_timestamp: string; expires_at: string
  checkout_url: string | null; transaction_id: string | null; paid_at: string | null
  snapshot: { order_number: string; property_id: number; property_name: string; property_address: string
    commencement_date: string; expiry_date: string; tenancy_months: number; tenant_name: string
    agreement_id: string; agreement_number: string; fees: { deposit: Money; service_fee: Money; tax: Money; current_total: Money } }
}
export interface PaymentResult extends PaymentResponse { property_image_url: string | null; booking_created_at: string; status_updated_at: string; failure_reason: string | null }
export type PaymentMethod = 'WECHAT_PAY' | 'ALIPAY' | 'CARD_CHECKOUT'
export interface PaymentMethodAvailability { method: PaymentMethod; available: boolean; test_mode: boolean; reason: string | null }
export interface TenantOrderItem {
  booking_id:number; order_id:string; agreement_id:string; agreement_number:string; property_id:number
  property_name:string; property_image_url:string|null; property_city:string; property_address:string
  lease_start_date:string|null; lease_end_date:string|null; lease_months:number|null
  settlement_currency:string; settlement_amount_minor:number; cny_reference_amount_minor:number
  property_currency:string; property_amount_minor:number; order_status:string; payment_status:string
  booking_status:string; status_label:string; created_at:string; expires_at:string
  remaining_payment_seconds:number; can_pay:boolean; payment_action_label:string|null; failure_reason:string|null
}
export interface TenantOrderDetail extends TenantOrderItem {
  applicant_name:string; applicant_phone_masked:string|null; applicant_email_masked:string|null
  property_type:string; property_country:string; property_description:string|null; monthly_rent_minor:number
  deposit_amount_minor:number; service_fee_amount_minor:number; tax_amount_minor:number
  exchange_rate:string; exchange_rate_source:string; exchange_rate_timestamp:string; status_updated_at:string
  paid_at:string|null; transaction_id_masked:string|null; webhook_confirmed:boolean; amounts_verified:boolean; inventory_reserved:boolean
}
export interface PaymentEligibility { booking_id:number; can_pay:boolean; order_status:string; payment_status:string; expires_at:string; reason:string|null; payment_id:string|null }

export const paymentService = {
  listMyOrders():Promise<TenantOrderItem[]> { return api.get('/payments/orders/my').then(r=>r.data.items) },
  getMyOrder(bookingId:number):Promise<TenantOrderDetail> { return api.get(`/payments/orders/my/${bookingId}`).then(r=>r.data) },
  validatePayment(bookingId:number):Promise<PaymentEligibility> { return api.post(`/payments/orders/my/${bookingId}/validate-payment`).then(r=>r.data) },
  createPayment(bookingId: number, idempotencyKey: string, paymentMethod: PaymentMethod = 'CARD_CHECKOUT'): Promise<PaymentResponse> {
    return api.post('/payments/create', { booking_id: bookingId, payment_method: paymentMethod }, { headers: { 'Idempotency-Key': idempotencyKey } }).then(r => r.data)
  },
  getAvailableMethods(): Promise<PaymentMethodAvailability[]> { return api.get('/payments/methods/availability').then(r => r.data) },
  getPayment(id: string): Promise<PaymentResponse> { return api.get(`/payments/${id}`).then(r => r.data) },
  getByBooking(id: number): Promise<PaymentResponse> { return api.get(`/payments/by-booking/${id}`).then(r => r.data) },
  getResult(id: number): Promise<PaymentResult> { return api.get(`/payments/result/by-booking/${id}`).then(r => r.data) },
}
